File: Web/packages/dom/test/scene-bus.test.ts
//
//  scene-bus.test.ts — the `scene` bus surface (dsx-game.md §2 G5): a mounted <scene>
//  element registers a handle into the kernel SceneRegistry seam, and the Core/Scene
//  module's web facet drives it through the MODULE CALL PATH. Under Node there is no
//  WebGL, so capture answers the typed capture_failed — the honest fallback — while
//  nodes/set/camera/pick/contacts/stats all answer from the corpus-pinned kernel folds.
//

import { test } from "node:test";
import assert from "node:assert/strict";

import type { XmlNode } from "@despia/compiler/xml";
import { sceneSurfaceSeam, sceneBusResolve, sceneBusEmit } from "@despia/kernel";
import { scene } from "../src/scene.ts";
import type { ElementApi } from "../src/elements.ts";
import type { MountCtx } from "../src/mount.ts";
import sceneFacet from "../../../../../ClosedSource/DSX/Modules/Core/Scene/web/index.js";

// ── the scene.test.ts DOM stand-in (each node --test file owns its process) ─────────

class FakeElement {
  tagName: string;
  className = "";
  textContent = "";
  attrs = new Map<string, string>();
  children: FakeElement[] = [];
  parent: FakeElement | null = null;
  constructor(tag: string) { this.tagName = tag.toUpperCase(); }
  setAttribute(name: string, value: string): void { this.attrs.set(name, value); }
  getAttribute(name: string): string | null { return this.attrs.get(name) ?? null; }
  appendChild(child: FakeElement): FakeElement { child.parent = this; this.children.push(child); return child; }
  remove(): void {
    if (this.parent === null) return;
    this.parent.children = this.parent.children.filter((c) => c !== this);
    this.parent = null;
  }
  addEventListener(): void {}
  getBoundingClientRect(): { width: number; height: number; left: number; top: number } {
    return { width: 320, height: 180, left: 0, top: 0 };
  }
  getContext(): null { return null; }   // no WebGL in Node — the honest fallback path
}

(globalThis as Record<string, unknown>)["document"] = {
  createElement: (tag: string) => new FakeElement(tag),
};

function xml(tag: string, attrs: { [k: string]: string } = {}, children: XmlNode[] = []): XmlNode {
  return { tag, attrs, children, text: "" };
}

function stubApi(node: XmlNode): ElementApi {
  return {
    bindText(expr, apply) { if (expr !== undefined) apply(expr.includes("{{") ? "" : expr); },
    bindValue() {},
    writeBack() {},
    handler() {},
    hasHandler(name) { return node.attrs[`on:${name}`] !== undefined; },
    children() {},
  };
}

type Failure = { code: string; message: string; recoverable: boolean };

function callCtx(values: Record<string, unknown> = {}): {
  ctx: { args(name: string): unknown; fail(code: string, message: string, recoverable: boolean, data?: unknown): void };
  failures: Failure[];
} {
  const failures: Failure[] = [];
  return {
    ctx: {
      args: (name) => values[name] ?? null,
      fail: (code, message, recoverable) => failures.push({ code, message, recoverable }),
    },
    failures,
  };
}

function mountScene(node: XmlNode): { host: FakeElement; dispose: () => void } {
  const disposers: Array<() => void> = [];
  const ctx = { disposers } as unknown as MountCtx;
  const host = scene(node, ctx, stubApi(node)) as unknown as FakeElement;
  return { host, dispose: () => disposers.forEach((d) => d()) };
}

function freshSeam(): void {
  const seam = sceneSurfaceSeam();
  seam.order.length = 0;
  seam.surfaces.clear();
  seam.listeners.clear();
}

const RIG = xml("scene", { id: "world", background: "#0b1020" }, [
  xml("camera", { position: "0 0 5", "look-at": "0 0 0" }),
  xml("light", { kind: "ambient", intensity: "0.5" }),
  xml("group", { id: "rig" }, [
    xml("box", { id: "crate", position: "-1 0 0", color: "#2563eb", collide: "box" }),
    xml("sphere", { id: "ball", position: "0 0 0", radius: "0.5", collide: "sphere" }),
  ]),
]);

test("a mounted scene registers into the seam under its id and unregisters on dispose", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  assert.ok(sceneBusResolve("world") !== null, "the id attr is the registry key");
  assert.equal(sceneBusResolve(null), sceneBusResolve("world"), "the only scene is the default target");
  mounted.dispose();
  assert.equal(sceneBusResolve("world"), null, "unmount unregisters");
  assert.equal(sceneBusResolve(null), null);
});

test("scene.nodes answers the resolved tree through the module face", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  const call = callCtx();
  const result = sceneFacet.actions.nodes(call.ctx) as { nodes: Array<{ kind: string; id: string; children: unknown[] }> };
  assert.deepEqual(call.failures, []);
  const kinds = result.nodes.map((n) => n.kind);
  assert.deepEqual(kinds, ["camera", "light", "group"]);
  const group = result.nodes[2]! as { kind: string; id: string; world: number[] | null; children: Array<{ id: string; kind: string; props: Record<string, string>; world: number[] | null }> };
  assert.equal(group.id, "rig");
  assert.deepEqual(group.children.map((c) => c.id), ["crate", "ball"]);
  assert.equal(group.children[0]!.props["position"], "-1 0 0");
  assert.equal(group.children[0]!.props["color"], "#2563eb");
  assert.deepEqual(group.children[0]!.world, [-1, 0, 0], "world position from the corpus fold");
  const one = sceneFacet.actions.nodes(callCtx({ id: "ball" }).ctx) as { nodes: Array<{ id: string }> };
  assert.equal(one.nodes.length, 1);
  assert.equal(one.nodes[0]!.id, "ball");
  const missing = callCtx({ id: "ghost" });
  sceneFacet.actions.nodes(missing.ctx);
  assert.equal(missing.failures[0]?.code, "node_not_found");
  mounted.dispose();
});

test("scene.set writes the resolved-attribute base plane; handlers refuse; unknown ids refuse", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  const ok = callCtx({ id: "ball", attr: "position", value: "0 2 0" });
  assert.deepEqual(sceneFacet.actions.set(ok.ctx), { ok: true });
  assert.deepEqual(ok.failures, []);
  const tree = (sceneFacet.actions.nodes(callCtx({ id: "ball" }).ctx) as { nodes: Array<{ props: Record<string, string>; world: number[] | null }> }).nodes[0]!;
  assert.equal(tree.props["position"], "0 2 0", "the write is visible on the resolved plane");
  assert.deepEqual(tree.world, [0, 2, 0], "the world fold sees the write");
  const bad = callCtx({ id: "ball", attr: "on:tap", value: "x()" });
  sceneFacet.actions.set(bad.ctx);
  assert.equal(bad.failures[0]?.code, "bad_attr");
  const ghost = callCtx({ id: "ghost", attr: "color", value: "#fff" });
  sceneFacet.actions.set(ghost.ctx);
  assert.equal(ghost.failures[0]?.code, "node_not_found");
  mounted.dispose();
});

test("scene.camera reads, writes, and refuses a malformed flyTo", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  const read = sceneFacet.actions.camera(callCtx().ctx) as { position: string; lookAt: string; fov: number; authored: boolean };
  assert.equal(read.position, "0 0 5");
  assert.equal(read.lookAt, "0 0 0");
  assert.equal(read.fov, 60);
  assert.equal(read.authored, true);
  const moved = sceneFacet.actions.camera(callCtx({ position: "0 1 8" }).ctx) as { position: string };
  assert.equal(moved.position, "0 1 8", "an immediate position write answers the new camera");
  const bad = callCtx({ flyTo: "not a triple" });
  sceneFacet.actions.camera(bad.ctx);
  assert.equal(bad.failures[0]?.code, "bad_value");
  // flyTo seeds the P5 transition plane: the BASE holds the destination immediately
  // (the glide is the render plane's business; Node has no rAF loop to advance it)
  const fly = callCtx({ flyTo: "0 0 12", durationMs: 200 });
  sceneFacet.actions.camera(fly.ctx);
  assert.deepEqual(fly.failures, []);
  mounted.dispose();
});

test("scene.pick answers the tap math without firing handlers; bad points refuse", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  // the ball sits at the origin, camera at 0 0 5 looking at it — dead center hits it
  const hit = sceneFacet.actions.pick(callCtx({ x: 0.5, y: 0.5 }).ctx) as { hit: boolean; id?: string; kind?: string };
  assert.equal(hit.hit, true);
  assert.equal(hit.id, "ball");
  assert.equal(hit.kind, "sphere");
  const missAnswer = sceneFacet.actions.pick(callCtx({ x: 0.02, y: 0.02 }).ctx) as { hit: boolean };
  assert.equal(missAnswer.hit, false, "a corner tap hits nothing");
  const bad = callCtx({ x: 2, y: 0.5 });
  sceneFacet.actions.pick(bad.ctx);
  assert.equal(bad.failures[0]?.code, "bad_point");
  mounted.dispose();
});

test("scene.contacts reads the currently overlapping collider pairs", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  const apart = sceneFacet.actions.contacts(callCtx().ctx) as { contacts: unknown[] };
  assert.deepEqual(apart.contacts, [], "crate at -1 and ball at 0 do not overlap");
  sceneFacet.actions.set(callCtx({ id: "crate", attr: "position", value: "0 0 0" }).ctx);
  const touching = sceneFacet.actions.contacts(callCtx().ctx) as { contacts: Array<{ a: string; b: string; depth: number }> };
  assert.equal(touching.contacts.length, 1);
  assert.deepEqual([touching.contacts[0]!.a, touching.contacts[0]!.b], ["crate", "ball"]);
  assert.ok(touching.contacts[0]!.depth > 0);
  mounted.dispose();
});

test("scene.stats answers the honest v0 profiler counters", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  const stats = sceneFacet.actions.stats(callCtx().ctx) as { nodes: number; animations: number; lastFrameDt: number; boundRows: number };
  assert.equal(stats.nodes, 5, "camera + light + group + box + sphere");
  assert.equal(stats.animations, 0);
  assert.equal(stats.lastFrameDt, 0);
  assert.equal(stats.boundRows, 0);
  mounted.dispose();
});

test("capture without a live renderer settles the typed capture_failed", () => {
  freshSeam();
  const mounted = mountScene(RIG);
  const call = callCtx();
  sceneFacet.actions.capture(call.ctx);
  assert.equal(call.failures[0]?.code, "capture_failed");
  assert.equal(call.failures[0]?.recoverable, true);
  mounted.dispose();
});

test("scene targeting: named scenes resolve, unknown targets settle scene_not_found", () => {
  freshSeam();
  const first = mountScene(RIG);
  const second = mountScene(xml("scene", { id: "hud" }, [xml("box", { id: "panel" })]));
  const named = sceneFacet.actions.stats(callCtx({ scene: "hud" }).ctx) as { nodes: number };
  assert.equal(named.nodes, 1);
  const fallback = sceneFacet.actions.stats(callCtx().ctx) as { nodes: number };
  assert.equal(fallback.nodes, 5, "no target = the FIRST mounted scene");
  const missing = callCtx({ scene: "nowhere" });
  sceneFacet.actions.stats(missing.ctx);
  assert.equal(missing.failures[0]?.code, "scene_not_found");
  second.dispose();
  first.dispose();
  const none = callCtx();
  sceneFacet.actions.stats(none.ctx);
  assert.equal(none.failures[0]?.code, "scene_not_found");
});

// ── G2 physics through the module face (dsx-game.md §2 G2): the write laws land on
// the solver via the SAME bus set action — never a second mutation path. Stepping
// itself is corpus-pinned (physics.json) and walk-verified; under Node there is no
// rAF loop, so these tests pin the write/read plumbing.

const PHYSICS_RIG = xml("scene", { id: "game", gravity: "0 -9.81 0" }, [
  xml("box", { id: "ground", physics: "static", position: "0 -0.5 0", size: "20 1 20" }),
  xml("sphere", { id: "ball", physics: "dynamic", position: "0 3 0", radius: "0.5" }),
  xml("box", { id: "hero", physics: "character", position: "2 0.9 2", size: "0.8 1.8 0.8", speed: "6" }),
]);

test("physics bodies expose solver state on the bus read (grounded/sleeping/velocity)", () => {
  freshSeam();
  const mounted = mountScene(PHYSICS_RIG);
  const tree = sceneFacet.actions.nodes(callCtx().ctx) as {
    nodes: Array<{ id: string; props: Record<string, string> }>;
  };
  const ball = tree.nodes.find((n) => n.id === "ball");
  assert.ok(ball !== undefined, "the dynamic body is on the read");
  assert.equal(ball!.props["grounded"], "false");
  assert.equal(ball!.props["sleeping"], "false");
  assert.equal(ball!.props["velocity"], "0 0 0");
  const ground = tree.nodes.find((n) => n.id === "ground");
  assert.equal(ground!.props["sleeping"], "false", "static bodies read too");
  mounted.dispose();
});

test("character intent and the impulse verb ride the bus set action (the G2 write laws)", () => {
  freshSeam();
  const mounted = mountScene(PHYSICS_RIG);
  // the move intent is attribute-shaped state: set writes the base plane the solver
  // reads each fixed step
  assert.deepEqual(sceneFacet.actions.set(callCtx({ id: "hero", attr: "move", value: "1 0" }).ctx), { ok: true });
  const hero = (sceneFacet.actions.nodes(callCtx({ id: "hero" }).ctx) as {
    nodes: Array<{ props: Record<string, string> }>;
  }).nodes[0]!;
  assert.equal(hero.props["move"], "1 0", "the intent landed on the resolved plane");
  // jump() IS a velocity write through the same set action (the impulse verb)
  assert.deepEqual(sceneFacet.actions.set(callCtx({ id: "hero", attr: "velocity", value: "0 8 0" }).ctx), { ok: true });
  const jumped = (sceneFacet.actions.nodes(callCtx({ id: "hero" }).ctx) as {
    nodes: Array<{ props: Record<string, string> }>;
  }).nodes[0]!;
  assert.equal(jumped.props["velocity"], "0 8 0", "the velocity write reached the solver body");
  // THE TELEPORT LAW: a dynamic body's position write moves it and RESETS velocity
  sceneFacet.actions.set(callCtx({ id: "ball", attr: "velocity", value: "1 2 3" }).ctx);
  sceneFacet.actions.set(callCtx({ id: "ball", attr: "position", value: "0 5 0" }).ctx);
  const ball = (sceneFacet.actions.nodes(callCtx({ id: "ball" }).ctx) as {
    nodes: Array<{ props: Record<string, string>; world: number[] | null }>;
  }).nodes[0]!;
  assert.equal(ball.props["velocity"], "0 0 0", "teleport reset the velocity");
  assert.deepEqual(ball.world, [0, 5, 0], "teleport moved the body (the override plane shows it)");
  mounted.dispose();
});

test("angular physics writes ride the bus and rotation lands on the render override plane", () => {
  freshSeam();
  const mounted = mountScene(PHYSICS_RIG);
  assert.deepEqual(
    sceneFacet.actions.set(callCtx({ id: "ball", attr: "rotation", value: "0 45 0" }).ctx),
    { ok: true },
  );
  sceneFacet.actions.set(callCtx({ id: "ball", attr: "angular-velocity", value: "0 3 0" }).ctx);
  sceneFacet.actions.set(callCtx({ id: "ball", attr: "torque", value: "0 2 0" }).ctx);
  let ball = (sceneFacet.actions.nodes(callCtx({ id: "ball" }).ctx) as {
    nodes: Array<{ props: Record<string, string> }>;
  }).nodes[0]!;
  assert.equal(ball.props["rotation"], "0 45 0", "rotation teleport is visible on the override plane");
  assert.equal(ball.props["angular-velocity"], "0 3 0", "angular velocity reached solver state");
  assert.equal(ball.props["torque"], "0 2 0", "persistent torque reached solver state");

  // A rotation teleport clears hidden spin, matching position teleport's no-glide law.
  sceneFacet.actions.set(callCtx({ id: "ball", attr: "rotation", value: "0 90 0" }).ctx);
  ball = (sceneFacet.actions.nodes(callCtx({ id: "ball" }).ctx) as {
    nodes: Array<{ props: Record<string, string> }>;
  }).nodes[0]!;
  assert.equal(ball.props["rotation"], "0 90 0");
  assert.equal(ball.props["angular-velocity"], "0 0 0", "rotation teleport reset angular velocity");
  mounted.dispose();
});

test("nested physics writes map local poses to root and rebuild without losing solver rotation", () => {
  freshSeam();
  const rig = xml("scene", { id: "nested" }, [
    xml("group", { id: "parent", position: "3 2 0", rotation: "0 0 20", scale: "2 1 1" }, [
      xml("box", { id: "child", physics: "dynamic", position: "1 0 0", rotation: "0 0 45", size: "2 1 1" }),
    ]),
  ]);
  const mounted = mountScene(rig);
  const child = (): { props: Record<string, string>; world: number[] | null } => {
    const tree = sceneFacet.actions.nodes(callCtx({ scene: "nested" }).ctx) as {
      nodes: Array<{ children: Array<{ props: Record<string, string>; world: number[] | null }> }>;
    };
    return tree.nodes[0]!.children[0]!;
  };
  let read = child();
  assert.equal(read.props["rotation"], "0 0 46.565051177078", "extraction exposes the root solver orientation");
  assert.deepEqual(read.world, [4.879385241571817, 2.6840402866513373, 0]);

  sceneFacet.actions.set(callCtx({ scene: "nested", id: "child", attr: "position", value: "2 0 0" }).ctx);
  sceneFacet.actions.set(callCtx({ scene: "nested", id: "child", attr: "rotation", value: "0 0 60" }).ctx);
  read = child();
  assert.deepEqual(read.world, [6.758770483143634, 3.3680805733026746, 0], "local teleport lifts through parent world");
  assert.equal(read.props["rotation"], "0 0 60.8933946491309", "local rotation composes into the root OBB frame");

  // Body-shaping writes rebuild extraction. The adapter must carry both public Euler
  // and the solver quaternion; this read catches a visible pose regression while the
  // subsequent kernel parity lane pins quaternion/contact continuity.
  sceneFacet.actions.set(callCtx({ scene: "nested", id: "child", attr: "mass", value: "2" }).ctx);
  read = child();
  assert.deepEqual(read.world, [6.758770483143634, 3.3680805733026746, 0]);
  assert.equal(read.props["rotation"], "0 0 60.8933946491309");
  mounted.dispose();
});

test("the seam's event door reaches the facet's boot subscription", () => {
  freshSeam();
  const events: Array<{ plane: string; kind: string; payload: unknown }> = [];
  sceneFacet.boot({
    delegate: { send: (event: string, payload: unknown) => events.push({ plane: "delegate", kind: event, payload }) },
    events: { publish: (name: string, payload: unknown) => events.push({ plane: "channel", kind: name, payload }) },
  } as never);
  sceneBusEmit("world", "ready", { scene: "world" });
  sceneBusEmit("world", "collide", { scene: "world", id: "a", other: "b", depth: 0.1 });
  sceneBusEmit("world", "internal", { scene: "world" });   // not a bus event — ignored
  assert.deepEqual(events.map((e) => `${e.plane}:${e.kind}`), [
    "delegate:scene.ready", "channel:scene:ready",
    "delegate:scene.collide", "channel:scene:collide",
  ]);
});

File: Engine/Android/core/src/test/kotlin/despia/engine/scene/SceneBusTest.kt
package despia.engine.scene

import despia.engine.StackXML
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * SceneRegistry.kt units — the `scene` BUS surface's :core half (dsx-game.md §2 G5):
 * the SceneRegistry seam (register/resolve/emit — the door the Core/Scene module
 * drives elements through) and the ONE shared SceneBusAdapter both JVM `<scene>`
 * elements construct (nodes · set · camera/flyTo · pick · contacts · stats). Every
 * markup fixture parses through StackXML.parse — the runtime's own parser, never a
 * second one; every number the adapter answers folds through the corpus-pinned kernel.
 */
class SceneBusTest {

    @AfterTest
    fun reset() {
        SceneRegistry.clearForTest()
    }

    private fun ir(markup: String): SceneIR {
        val root = StackXML.parse(markup)
        assertNotNull(root, "fixture markup must parse")
        return parseScene(root, diag = { d -> error("unexpected parse diagnostic: ${d.message}") })
    }

    private fun adapter(markup: String, vars: MutableMap<String, Any?> = HashMap()): SceneBusAdapter {
        val parsed = ir(markup)
        val animator = SceneAnimator()
        val reconciler = SceneBindReconciler(evalBind = { expr, _ -> vars[expr] }, animator = animator)
        reconciler.register(parsed.nodes, null)
        val bus = SceneBusAdapter(parsed, animator, reconciler)
        bus.rawBase = { _, _, raw -> interpolateSceneHoles(raw) { expr -> vars[expr] } }
        animator.attach(parsed.nodes, null, bus.resolveBase)
        return bus
    }

    private val RIG = """
        <scene id="world" background="#0b1020">
          <camera position="0 0 5" look-at="0 0 0"/>
          <light kind="ambient" intensity="0.5"/>
          <group id="rig">
            <box id="crate" position="-1 0 0" color="#2563eb" collide="box"/>
            <sphere id="ball" position="0 0 0" radius="0.5" collide="sphere"/>
          </group>
        </scene>
    """.trimIndent()

    // ── the registry seam ────────────────────────────────────────────────────────────

    @Test
    fun registryKeysResolveAndUnregister() {
        val a = adapter(RIG)
        val key = SceneRegistry.register(a)
        assertEquals("world", key, "the scene root's id attr is the registry key")
        assertEquals(a, SceneRegistry.resolve("world"))
        assertEquals(a, SceneRegistry.resolve(null), "absent target = the FIRST mounted scene")
        val b = adapter("<scene><box id=\"panel\"/></scene>")
        val bKey = SceneRegistry.register(b)
        assertTrue(bKey.startsWith("scene#"), "an id-less scene takes the auto key, saw $bKey")
        assertEquals(a, SceneRegistry.resolve(null), "the first mount stays the default")
        SceneRegistry.unregister(key)
        assertNull(SceneRegistry.resolve("world"), "unregister removes the key")
        assertEquals(b, SceneRegistry.resolve(null), "the next mounted scene becomes the default")
    }

    @Test
    fun duplicateIdsStayAddressable() {
        val a = adapter(RIG)
        val b = adapter(RIG)
        assertEquals("world", SceneRegistry.register(a))
        val second = SceneRegistry.register(b)
        assertTrue(second.startsWith("scene#"), "a duplicate id falls back to the auto key, saw $second")
        assertEquals(a, SceneRegistry.resolve("world"))
        assertEquals(b, SceneRegistry.resolve(second))
    }

    @Test
    fun emitReachesListenersAndSurvivesAThrowingOne() {
        val seen = ArrayList<String>()
        val off = SceneRegistry.listen { scene, kind, _ -> error("bad observer $scene.$kind") }
        SceneRegistry.listen { scene, kind, payload -> seen.add("$scene.$kind:${payload["id"]}") }
        SceneRegistry.emit("world", "collide", mapOf("id" to "crate"))
        assertEquals(listOf("world.collide:crate"), seen, "the throwing listener never blocks the fan-out")
        off()
    }

    // ── the adapter: nodes ───────────────────────────────────────────────────────────

    @Test
    fun nodesAnswersTheResolvedTree() {
        val a = adapter(RIG)
        val tree = a.nodes()
        assertEquals(listOf("camera", "light", "group"), tree.map { it.kind })
        val group = tree[2]
        assertEquals("rig", group.id)
        assertEquals(listOf("crate", "ball"), group.children.map { it.id })
        val crate = group.children[0]
        assertEquals("-1 0 0", crate.props["position"], "resolved authored position")
        assertEquals("#2563eb", crate.props["color"])
        val world = assertNotNull(crate.world, "a box carries a world entry")
        assertEquals(-1.0, world[0], 1e-9)
        assertEquals(0.0, world[1], 1e-9)
        assertEquals(0.0, world[2], 1e-9)
    }

    // ── set: the base overlay + the P5 glide plane ───────────────────────────────────

    @Test
    fun setWritesTheBasePlaneAndValidates() {
        val a = adapter(RIG)
        assertEquals("bad_attr", a.set("ball", "on:tap", "x()"))
        assertEquals("bad_attr", a.set("ball", "id", "other"))
        assertEquals("node_not_found", a.set("ghost", "color", "#fff"))
        var invalidated = 0
        a.invalidate = { invalidated += 1 }
        assertEquals("ok", a.set("ball", "position", "0 2 0"))
        assertEquals(1, invalidated, "a bus set invalidates the element")
        val ball = a.nodes()[2].children[1]
        assertEquals("0 2 0", ball.props["position"], "the write is visible on the resolved plane")
        assertEquals(2.0, assertNotNull(ball.world)[1], 1e-9, "the world fold sees the write")
    }

    @Test
    fun setOnATransitionCoveredPropertyGlides() {
        val a = adapter(
            """
            <scene>
              <camera position="0 0 5"/>
              <box id="crate" position="0 0 0" transition="position 200ms linear"/>
            </scene>
            """.trimIndent(),
        )
        val animator = run {
            // drive the SAME animator the adapter writes through: read it back via stats
            assertEquals(0, a.stats().animations, "no animation before the write")
            assertEquals("ok", a.set("crate", "position", "4 0 0"))
            a
        }
        assertEquals(1, animator.stats().animations,
            "a set on a transition-covered property starts the P5 glide")
    }

    // ── camera: read, write, flyTo ───────────────────────────────────────────────────

    @Test
    fun cameraReadsWritesAndFliesThroughTheP5Plane() {
        val a = adapter(RIG)
        val read = a.camera()
        assertEquals("0 0 5", read.position)
        assertEquals("0 0 0", read.lookAt)
        assertEquals(60.0, read.fov, 1e-9)
        assertTrue(read.authored)
        assertEquals("bad_value", a.cameraSet(null, null, "not a triple", null))
        assertEquals("ok", a.cameraSet("0 1 8", null, null, null))
        assertEquals("0 1 8", a.camera().position, "an immediate write answers the new camera")
        assertEquals("ok", a.cameraSet(null, null, "0 0 12", 200.0))
        assertEquals(1, a.stats().animations, "flyTo seeds one glide on the transition plane")
        // the glide starts from the CURRENT rendered position (never snap) …
        assertEquals("0 1 8", a.camera().position, "the override holds the pre-fly position at t=0")
    }

    @Test
    fun cameraLessSceneReadsDefaultsAndRefusesWrites() {
        val a = adapter("<scene><box id=\"panel\"/></scene>")
        val read = a.camera()
        assertEquals("0 0 5", read.position, "the corpus camera default")
        assertEquals(false, read.authored)
        assertEquals("no_camera", a.cameraSet("0 1 8", null, null, null))
    }

    // ── pick: the tap math, no handlers fired ────────────────────────────────────────

    @Test
    fun pickAnswersTheNearestHitWithoutHandlers() {
        val a = adapter(RIG)
        a.viewWidth = 320
        a.viewHeight = 180
        val hit = assertNotNull(a.pick(0.5, 0.5), "dead center hits the origin ball")
        assertEquals("ball", hit.id)
        assertEquals("sphere", hit.kind)
        assertNull(a.pick(0.02, 0.02), "a corner tap hits nothing")
    }

    // ── contacts: the current overlap read ───────────────────────────────────────────

    @Test
    fun contactsReadsCurrentOverlaps() {
        val a = adapter(RIG)
        assertEquals(emptyList(), a.contacts().map { it.a }, "crate at -1 and ball at 0 do not overlap")
        assertEquals("ok", a.set("crate", "position", "0 0 0"))
        val touching = a.contacts()
        assertEquals(1, touching.size)
        assertEquals("crate", touching[0].a)
        assertEquals("ball", touching[0].b)
        assertTrue(touching[0].depth > 0.0, "an overlap carries a positive depth, saw ${touching[0].depth}")
    }

    // ── stats: the honest v0 profiler read ───────────────────────────────────────────

    @Test
    fun statsCountsNodesAnimationsAndBoundRows() {
        val vars = HashMap<String, Any?>()
        vars["rows"] = listOf(mapOf("id" to "e1"), mapOf("id" to "e2"))
        val parsed = ir(
            """
            <scene>
              <camera position="0 0 5"/>
              <group bind="rows" key="id"><box position="0 0 0"/></group>
            </scene>
            """.trimIndent(),
        )
        val animator = SceneAnimator()
        val reconciler = SceneBindReconciler(evalBind = { expr, _ -> vars[expr.trim()] }, animator = animator)
        reconciler.register(parsed.nodes, null)
        val bus = SceneBusAdapter(parsed, animator, reconciler)
        bus.rawBase = { _, _, raw -> interpolateSceneHoles(raw) { expr -> vars[expr] } }
        animator.attach(parsed.nodes, null, bus.resolveBase)
        reconciler.reconcile(bus.resolveBase)
        val stats = bus.stats()
        assertEquals(2, stats.boundRows, "two live rows")
        assertEquals(4, stats.nodes, "camera + group + two instantiated boxes")
        assertEquals(0, stats.animations)
        assertEquals(0.0, stats.lastFrameDt, 1e-9)
        bus.lastFrameDt = 16.7
        assertEquals(16.7, bus.stats().lastFrameDt, 1e-9, "the element stamps each emitted tick")
    }

    // ── capture: element-owned encoder seam ──────────────────────────────────────────

    @Test
    fun captureAnswersNullUntilAnElementBindsItsEncoder() {
        val a = adapter(RIG)
        assertNull(a.capture(), "no encoder bound = the module's typed capture_failed")
        a.captureFrame = { SceneBusCapture("QUJD", 320, 180) }
        val captured = assertNotNull(a.capture())
        assertEquals("QUJD", captured.image)
        assertEquals(320, captured.width)
        assertEquals(180, captured.height)
    }

    // ── G6: a bus write of the SPAWN pose is still a COMMAND (the 2D-walk finding) ───

    @Test
    fun aBusPositionWriteOfTheUNCHANGEDSpawnStringStillTeleportsASolverOwnedBody() {
        val parsed = ir(
            """
            <scene id="drop" mode="2d" gravity="0 -9.81 0">
              <sprite id="crate" physics="dynamic" position="0 3.5 0.5" size="1 1"/>
            </scene>
            """.trimIndent(),
        )
        val animator = SceneAnimator()
        val reconciler = SceneBindReconciler(evalBind = { _, _ -> null }, animator = animator)
        reconciler.register(parsed.nodes, null)
        val bus = SceneBusAdapter(parsed, animator, reconciler)
        bus.rawBase = { _, _, raw -> raw }
        animator.attach(parsed.nodes, null, bus.resolveBase)
        val physics = ScenePhysicsRuntime(parsed, animator, diag = null)
        bus.physicsInfo = { n -> physics.info(n) }
        bus.forcePhysicsWrite = { n, attr -> physics.forceWrite(n, attr) }
        physics.noteBase(bus.resolveBase)

        // fall for a second — the solver owns the rendered position, the BASE cache
        // still holds the authored spawn string "0 3.5 0.5"
        val render = animator.resolve(bus.resolveBase)
        repeat(60) { physics.advance(1.0 / 60.0, render, bus.resolveBase) }
        val node = findSceneNode(parsed.nodes, "crate")!!
        val fell = resolvedProps(node, animator.resolve(bus.resolveBase))
        assertTrue(fell.position[1] < 3.0, "the 2D crate fell (y ${fell.position[1]})")
        assertEquals(0.5, fell.position[2], "the z-lock held z at its spawn layer")

        // writing the SAME spawn string back is "put it where it started" — it must
        // teleport, not dedupe away
        assertEquals("ok", bus.set("crate", "position", "0 3.5 0.5"))
        physics.noteBase(bus.resolveBase)   // the element's invalidate → the next scan
        val reset = resolvedProps(node, animator.resolve(bus.resolveBase))
        assertEquals(3.5, reset.position[1], 1e-9, "the spawn-pose write teleported the body back")
        assertEquals(0.5, reset.position[2], "…and re-anchored on the same z layer")
    }
}
